File: educative/js/find_substring.js
function find_substring(str, pattern) {
  let startIndex = 0;
  let start = 0;
  let res = "";
  let endIndex = 0;
  const charsFrequency = {};
  for (let i = 0; i < pattern.length; i++) {
    const char = pattern[i];
    if (!(char in charsFrequency)) {
      charsFrequency[char] = 0;
    }
    charsFrequency[char]++;
  }

  let minlen = str.length + 1;
  let matches = 0;
  for (let end = 0; end < str.length; end++) {
    const rightchar = str[end];
    if (rightchar in charsFrequency) {
      charsFrequency[rightchar]--;
      if (charsFrequency[rightchar] === 0) {
        matches++;
      }
    }
    while (matches === pattern.length) {
      if (minlen > end - start + 1) {
        minlen = end - start + 1;
        startIndex = start;
        endIndex = end;
      }
      const leftchar = str[start];
      start++;
      if (leftchar in charsFrequency) {
        if (charsFrequency[leftchar] === 0) {
          matches--;
        }
        charsFrequency[leftchar]++;
      }
    }
  }

  if (minlen > str.length) {
    return "";
  }

  for (startIndex; startIndex <= endIndex; startIndex++) {
    res += str[startIndex];
  }
  return res;
  // return str.substring(startIndex, startIndex + minlen);
}

console.log(find_substring("aabdec", "abc"));
console.log(find_substring("abdbca", "abc"));
console.log(find_substring("adcad", "abc"));

File: educative/python/SlidingWindowPattern/find_substring.py



def find_substring(str, pattern):
  res = ""
  endIndex = 0
  startIndex = 0
  charmap = {}
  minlen = len(str) + 1
  start = 0
  matched = 0


  for chr in pattern:
    if chr not in charmap:
      charmap[chr] = 0
    charmap[chr] += 1

  for end, _ in enumerate(str):
    rightchar = str[end]
    if rightchar in charmap:
      charmap[rightchar] -=1
      if charmap[rightchar] >= 0:
        matched+=1

    while matched == len(pattern):
      if minlen > (end - start + 1):
        minlen = (end - start + 1)
        endIndex = end
        startIndex = start
      leftchar = str[start]
      start+=1

      if leftchar in charmap:
        if charmap[leftchar] == 0:
          matched -=1
        charmap[leftchar] +=1

  if minlen > len(str):
    return ""
    
    
  while startIndex <= endIndex:
    res += str[startIndex]
    startIndex+=1
  return  res
      






      

  






print(find_substring("aabdec", "abc"));
print(find_substring("abdbca", "abc"));
print(find_substring("adcad", "abc"));